command/agent/rpc_client.go

```diff
@@ -345,6 +346,161 @@ func (c *RPCClient) Monitor(level logutils.LogLevel, ch chan<- string) (StreamHa
 	}
 }
 
+// NodeResponse is used to return the response of a query
+type NodeResponse struct {
+	From    string
+	Payload []byte
+}
+
+type queryHandler struct {
+	client *RPCClient
+	closed bool
+	init   bool
+	initCh chan<- error
+	ackCh  chan<- string
+	respCh chan<- NodeResponse
+	seq    uint64
+}
+
+func (qh *queryHandler) Handle(resp *responseHeader) {
+	log.Printf("[DEBUG] jfs: consul/command/agent/rpc_client.go Handle() invoked...")
+	// Initialize on the first response
+	if !qh.init {
+		qh.init = true
+		qh.initCh <- strToError(resp.Error)
+		return
+	}
+
+	// Decode the query response
+	var rec queryRecord
+	if err := qh.client.dec.Decode(&rec); err != nil {
+		log.Printf("[ERR] Failed to decode query response: %v", err)
+		qh.client.deregisterHandler(qh.seq)
+		return
+	}
+
+	switch rec.Type {
+	case queryRecordAck:
+		select {
+		case qh.ackCh <- rec.From:
+		default:
+			log.Printf("[ERR] Dropping query ack, channel full")
+		}
+
+	case queryRecordResponse:
+		select {
+		case qh.respCh <- NodeResponse{rec.From, rec.Payload}:
+		default:
+			log.Printf("[ERR] Dropping query response, channel full")
+		}
+
+	case queryRecordDone:
+		// No further records coming
+		qh.client.deregisterHandler(qh.seq)
+
+	default:
+		log.Printf("[ERR] Unrecognized query record type: %s", rec.Type)
+	}
+}
+
+func (qh *queryHandler) Cleanup() {
+	log.Printf("[DEBUG] jfs: consul/command/agent/rpc_client.go Cleanup() invoked...")
+	if !qh.closed {
+		if !qh.init {
+			qh.init = true
+			qh.initCh <- fmt.Errorf("Stream closed")
+		}
+		if qh.ackCh != nil {
+			close(qh.ackCh)
+		}
+		if qh.respCh != nil {
+			close(qh.respCh)
+		}
+		qh.closed = true
+	}
+}
+
+const (
+	queryRecordAck      = "ack"
+	queryRecordResponse = "response"
+	queryRecordDone     = "done"
+)
+
+type queryRecord struct {
+	Type    string
+	From    string
+	Payload []byte
+}
+
+type queryRequest struct {
+	FilterNodes []string
+	FilterTags  map[string]string
+	RequestAck  bool
+	Timeout     time.Duration
+	Name        string
+	Payload     []byte
+}
+
+// QueryParam is provided to query set various settings.
```

Comment:
This comment was directly copied from the serf project's implementation of the 'reachability' command.  I believe the intended comment is:
// QueryParam is provided to set various query settings
Or perhaps more clearly:
// QueryParam is provided to customize various query settings